test/OrchardCore.Tests/Modules/OrchardCore.AutoSetup/AutoSetupMiddlewareTests.cs

```diff
@@ -0,0 +1,161 @@
+        _mockAutoSetupService.Verify(s => s.SetupTenantAsync(It.IsAny<TenantSetupOptions>(), It.IsAny<ShellSettings>()), Times.Never);
+    }
+
+    [Fact]
+    public async Task InvokeAsync_FailedSetup_ReturnsServiceUnavailable()
+    {
+        // Arrange
+        _shellSettings.State = TenantState.Uninitialized;
+
+        SetupDistributedLockMock(true);
+
+        var setupContext = new SetupContext { Errors = new Dictionary<string, string> { { "Error", "Test error" } } };
+        _mockAutoSetupService.Setup(s => s.SetupTenantAsync(It.IsAny<TenantSetupOptions>(), It.IsAny<ShellSettings>()))
+            .ReturnsAsync((setupContext, false));
+
+        var httpContext = new DefaultHttpContext();
+        httpContext.RequestServices = new ServiceCollection()
+            .AddSingleton(_mockAutoSetupService.Object)
+            .BuildServiceProvider();
+
+        var middleware = new AutoSetupMiddleware(
+            next: (innerHttpContext) => Task.CompletedTask,
+            _mockShellHost.Object,
+            _shellSettings,
+            _mockShellSettingsManager.Object,
+            _mockDistributedLock.Object,
+            _mockOptions.Object);
+
+        // Act
+        await middleware.InvokeAsync(httpContext);
+
+        // Assert
+        Assert.Equal(StatusCodes.Status503ServiceUnavailable, httpContext.Response.StatusCode);
+    }
+
+    [Fact]
+    public async Task InvokeAsync_UnInitializedShell_PerformsSetup()
+    {
+        // Arrange
+        _shellSettings.State = TenantState.Uninitialized;
+
+        SetupDistributedLockMock(true);
+
+        var setupContext = new SetupContext();
+        _mockAutoSetupService.Setup(s => s.SetupTenantAsync(It.IsAny<TenantSetupOptions>(), It.IsAny<ShellSettings>()))
+            .ReturnsAsync((setupContext, true));
+
+        var httpContext = new DefaultHttpContext();
+        httpContext.RequestServices = new ServiceCollection()
+            .AddSingleton(_mockAutoSetupService.Object)
+            .BuildServiceProvider();
+
+        var middleware = new AutoSetupMiddleware(
+            next: (innerHttpContext) => Task.CompletedTask,
+            _mockShellHost.Object,
+            _shellSettings,
+            _mockShellSettingsManager.Object,
+            _mockDistributedLock.Object,
+            _mockOptions.Object);
+
+        // Act
+        await middleware.InvokeAsync(httpContext);
+
+        // Assert
+        Assert.Equal(StatusCodes.Status302Found, httpContext.Response.StatusCode); // Redirect
+        _mockAutoSetupService.Verify(s => s.SetupTenantAsync(It.IsAny<TenantSetupOptions>(), It.IsAny<ShellSettings>()), Times.Once);
+    }
+
+    [Fact]
+    public async Task InvokeAsync_FailedLockAcquisition_ThrowsTimeoutException()
+    {
+        // Arrange
+        _shellSettings.State = TenantState.Uninitialized;
+
+        SetupDistributedLockMock(false);
+
+        var httpContext = new DefaultHttpContext();
+
+        var middleware = new AutoSetupMiddleware(
```

Comment:
You can create a `private` method for middleware creation to make reusable across the unit tests